service/history/timerQueueStandbyTaskExecutor.go

```diff
@@ -202,8 +202,14 @@ func (t *timerQueueStandbyTaskExecutor) executeActivityTimeoutTask(
 		if err != nil {
 			return nil, err
 		}
+
+		// NOTE: LastHeartbeatTimeoutVisibility is for deduping heartbeat timer creation as it's possible
+		// one heartbeat task was persisted multiple times with different taskIDs due to the retry logic
+		// for updating workflow execution. In that case, only one new heartbeat timeout task should be
+		// created.
 		isHeartBeatTask := timerTask.TimeoutType == int(workflow.TimeoutTypeHeartbeat)
-		if activityInfo, ok := mutableState.GetActivityInfo(timerTask.EventID); isHeartBeatTask && ok {
+		activityInfo, ok := mutableState.GetActivityInfo(timerTask.EventID)
+		if isHeartBeatTask && ok && activityInfo.LastHeartbeatTimeoutVisibility <= timerTask.VisibilityTimestamp.Unix() {
```

Comment:
`LastHeartbeatTimeoutVisibility` you may want to rename this var to indicate this is in seconds

Reply:
Renamed to `LastHeartbeatTimeoutVisibilityInSeconds`